compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -656,7 +656,16 @@ pub enum TerminatorKind<'tcx> {
     /// The `replace` flag indicates whether this terminator was created as part of an assignment.
     /// This should only be used for diagnostic purposes, and does not have any operational
     /// meaning.
-    Drop { place: Place<'tcx>, target: BasicBlock, unwind: UnwindAction, replace: bool },
+    Drop {
+        place: Place<'tcx>,
+        target: BasicBlock,
+        unwind: UnwindAction,
+        replace: bool,
+        /// Cleanup to be done if the coroutine is dropped at this suspend point (for async drop).
```

Comment:
What does it mean for the coroutine to be "dropped at this suspend point"? I don't understand under which conditions the code will jump to the `drop` block.

Does it really make sense to make the existing `Drop` terminator do *both* regular drop and async drop? Either way, the doc comment for the terminator needs to be a self-contained explanation of what exactly happens when the terminator is executed.

Reply:
Async drop in StateTransform pass will be expanded into yield point for poll-loop of internal drop coroutine. So, we need the same `drop` target block prepared for async Drop, as for Yield terminator. It means - "where the main coroutine should go, if it is dropped at this yield point (drop pool-loop)".
StateTransform pass generates "coroutine_resume" function for poll and "coroutine_drop_async" for drop of coroutine itself. Both functions are switches: for `resume` function state target blocks are taken from `target` field of Yield (or async Drop) and for `drop` function state target blocks are taken from `drop` field of Yield (or async Drop).

Yes, I think keeping Drop terminator for both sync/async is good atm because async Drop is optionally expanded if conditions are met. For sync context (just inside normal function) and for unwind - we can't use AsyncDrop atm, so sync drop must be as a fallback.

Drop terminator comments updated.